src/internal/components/dnd-container/index.tsx

```diff
@@ -1,61 +1,39 @@
 // SPDX-License-Identifier: Apache-2.0
 
-import React from 'react';
-import { DndContext, DraggableAttributes, DragOverlay } from '@dnd-kit/core';
-import { SyntheticListenerMap } from '@dnd-kit/core/dist/hooks/utilities';
+import React, { useMemo } from 'react';
+import { DndContext, DragOverlay } from '@dnd-kit/core';
 import { arrayMove, SortableContext, useSortable, verticalListSortingStrategy } from '@dnd-kit/sortable';
 import { CSS } from '@dnd-kit/utilities';
 
+import { joinStrings } from '../../utils/strings';
 import Portal from '../portal';
-import { ReorderAnnouncements, ReorderOptions } from './interfaces';
+import { DndContainerItem, DndContainerProps, RenderItemProps } from './interfaces';
 import useDragAndDropReorder from './use-drag-and-drop-reorder';
 import useLiveAnnouncements from './use-live-announcements';
 
 import styles from './styles.css.js';
 
-export interface DndContainerProps<Option> extends ReorderOptions<Option> {
-  onChange: (sortedOption: readonly Option[]) => void;
-  disabled?: boolean;
-  renderOption: (props: RenderOptionProps<Option>) => React.ReactNode;
-  renderActiveOption: (props: RenderActiveOptionProps<Option>) => React.ReactNode;
-  i18nStrings: ReorderAnnouncements;
-  dragOverlayClassName?: string;
-  children?: (optionsContent: React.ReactNode) => React.ReactNode;
-}
-
-interface RenderOptionProps<Option> extends RenderActiveOptionProps<Option> {
-  ref: React.RefCallback<HTMLElement>;
-  style: React.CSSProperties;
-  isDragging: boolean;
-  isSorting: boolean;
-  attributes: DraggableAttributes;
-}
-
-interface RenderActiveOptionProps<Option> {
-  option: Option;
-  dragHandleAriaLabel?: string;
-  listeners?: SyntheticListenerMap;
-}
-
-export function DndContainer<Option>({
-  getId,
-  sortedOptions,
-  onChange,
-  disabled,
-  renderOption,
-  renderActiveOption,
+export function DndContainer<Data>({
+  items,
+  renderItem,
+  onItemsChange,
+  disableReorder,
   i18nStrings,
   dragOverlayClassName = styles['drag-overlay'],
-  children = content => content,
-}: DndContainerProps<Option>) {
-  const { activeItem, collisionDetection, handleKeyDown, sensors, setActiveItem } = useDragAndDropReorder({
-    sortedOptions,
-    getId,
+}: DndContainerProps<Data>) {
+  const { activeItemId, setActiveItemId, collisionDetection, handleKeyDown, sensors } = useDragAndDropReorder({
+    items,
   });
-  const activeOption = activeItem ? sortedOptions.find(option => getId(option) === activeItem) : null;
-  const isDragging = activeItem !== null;
-  const announcements = useLiveAnnouncements({ sortedOptions, getId, isDragging, ...i18nStrings });
+  const activeItem = activeItemId ? items.find(item => item.id === activeItemId) : null;
+  const isDragging = activeItemId !== null;
+  const announcements = useLiveAnnouncements({ items, isDragging, ...i18nStrings });
+  const accessibilityContainer = useMemo(() => document.createElement('div'), []);
```

Comment:
By default, the DndContext renders accessibility nodes (for ARIA description and announcements) in-place. That creates an HTML violation when using the context for semantic elements such as lists or tables (this is an existing issue with collection prefs).

To address the issue, the previous implementation used render-prop children. The current implementation uses the accessibility.container prop of the DndContext that specifies location where the a11y props must go. The accessibility.container takes an Element which does not exist during the first render unless provided from the outside. To solve this, we create the element with document.createElement() and append it to the portal node when the portal becomes available.